File: src/sampler.ts
import fs from 'node:fs';
import { loadSuiteInstances, OFFICIAL_META } from './dataset';
import * as difficulty from './difficulty';
import { Instance, SuiteLevel, SuiteManifest } from './schemas';
import { FIXED_SUITES_PATH } from './paths';

export const QUOTAS: Record<string, Record<string, number>> = {
  smoke6: { easy: 2, medium: 2, hard: 2 },
  core12: { easy: 2, medium: 6, hard: 4 },
  confirm24_addon: { easy: 2, medium: 6, hard: 4 },
};

export const REPO_MAX: Record<string, number> = {
  smoke6: 1,
  core12: 2,
  confirm24: 2,
};

export const TYPE_FLOOR: Record<string, number> = {
  bug: 3,
  feature: 3,
  refactor: 2,
  infra: 2,
};

function diversityGain(inst: Instance, selected: Instance[]): number {
  const dims: Array<keyof Instance> = [
    'language_family',
    'task_type',
    'repo',
    'knowledge_domain',
  ];
  let fresh = 0;
  for (const dim of dims) {
    const covered = new Set(selected.map((s) => String(s[dim])));
    if (!covered.has(String(inst[dim]))) fresh += 1;
  }
  return fresh / dims.length;
}

function selectBand(
  candidates: Instance[],
  band: string,
  selected: Instance[],
  repoCount: Record<string, number>,
  repoMax: number,
  typeNeeds: Record<string, number>,
): { inst: Instance | null; relaxations: string[] } {
  const relaxations: string[] = [];
  const pool = candidates.filter((c) => difficulty.difficultyBand(c) === band);

  function eligible(c: Instance, relaxRepo: boolean): boolean {
    if ((repoCount[c.repo] ?? 0) >= repoMax && !relaxRepo) return false;
    return true;
  }

  function score(c: Instance): number {
    const bandP = difficulty.bandPassProb(band);
    const typeBonus = (typeNeeds[c.task_type] ?? 0) > 0 ? 0.15 : 0.0;
    return difficulty.infoScore(c, bandP, diversityGain(c, selected)) + typeBonus;
  }

  for (const relaxRepo of [false, true]) {
    const avail = pool.filter((c) => eligible(c, relaxRepo));
    if (relaxRepo && avail.length > 0) {
      relaxations.push(`${band}: 仓库上限约束放宽(候选不足)`);
    }
    if (avail.length > 0) {
      let best = avail[0]!;
      for (const c of avail) {
        const sc = score(c);
        const bestSc = score(best);
        if (sc > bestSc || (sc === bestSc && c.instance_id > best.instance_id)) {
          best = c;
        }
      }
      return { inst: best, relaxations };
    }
  }
  return { inst: null, relaxations };
}

function fillQuotas(
  candidates: Instance[],
  quotas: Record<string, number>,
  selected: Instance[],
  repoCount: Record<string, number>,
  repoMax: number,
  baseSelected: Instance[] = [],
  baseRepoCount: Record<string, number> = {},
): string[] {
  const relaxations: string[] = [];
  const takenIds = new Set(selected.map((i) => i.instance_id));

  function pickBand(band: string): Instance | null {
    const typeNeeds: Record<string, number> = { ...TYPE_FLOOR };
    for (const i of selected) {
      if (typeNeeds[i.task_type] != null && typeNeeds[i.task_type] > 0) {
        typeNeeds[i.task_type] -= 1;
      }
    }
    const { inst, relaxations: rel } = selectBand(
      candidates.filter((c) => !takenIds.has(c.instance_id)),
      band,
      [...baseSelected, ...selected],
      { ...baseRepoCount, ...repoCount },
      repoMax,
      typeNeeds,
    );
    relaxations.push(...rel);
    return inst;
  }

  const remaining: Record<string, number> = { ...quotas };
  const driftOrder: Record<string, string[]> = {
    easy: ['medium'],
    medium: ['easy', 'hard'],
    hard: ['medium'],
  };

  while (Object.values(remaining).some((v) => v > 0)) {
    const bands = Object.keys(remaining).filter((b) => remaining[b]! > 0);
    let band = bands[0]!;
    for (const b of bands) {
      if (
        remaining[b]! > remaining[band]! ||
        (remaining[b]! === remaining[band]! && b > band)
      ) {
        band = b;
      }
    }

    let inst = pickBand(band);
    if (inst == null) {
      for (const alt of driftOrder[band] ?? []) {
        if ((remaining[alt] ?? 0) > 0) continue;
        inst = pickBand(alt);
        if (inst != null) {
          relaxations.push(
            `${band} 候选池耗尽,由相邻带 ${alt} 漂移补位(${inst.instance_id})`,
          );
          break;
        }
      }
      if (inst == null) {
        relaxations.push(
          `${band} 带配额缺口 ${remaining[band]} 无法满足(候选池不足)`,
        );
        remaining[band] = 0;
        continue;
      }
    }
    selected.push(inst);
    takenIds.add(inst.instance_id);
    repoCount[inst.repo] = (repoCount[inst.repo] ?? 0) + 1;
    remaining[band]! -= 1;
  }
  return relaxations;
}

export interface FixedSuiteDefinition {
  suite_id: string;
  name: string;
  level: SuiteLevel;
  description?: string;
  instance_ids: string[];
}

export function loadFixedSuiteDefinitions(): FixedSuiteDefinition[] {
  const raw = JSON.parse(fs.readFileSync(FIXED_SUITES_PATH, 'utf-8')) as {
    suites?: Array<Record<string, unknown>>;
  };
  if (!Array.isArray(raw.suites)) {
    throw new Error(`fixed_suites.json must contain a "suites" array`);
  }
  return raw.suites.map((s) => ({
    suite_id: String(s.suite_id ?? ''),
    name: String(s.name ?? s.suite_id ?? ''),
    level: String(s.level ?? 'smoke6') as SuiteLevel,
    description: s.description ? String(s.description) : '',
    instance_ids: Array.isArray(s.instance_ids) ? s.instance_ids.map(String) : [],
  }));
}

function fixedSuiteMetaVersion(): string {
  try {
    const raw = JSON.parse(fs.readFileSync(FIXED_SUITES_PATH, 'utf-8')) as {
      _meta?: { version?: string };
    };
    return String(raw._meta?.version ?? 'official-v1');
  } catch {
    return 'official-v1';
  }
}

export function listFixedSuites(): Array<Record<string, unknown>> {
  const version = fixedSuiteMetaVersion();
  return loadFixedSuiteDefinitions().map((d) => ({
    suite_id: d.suite_id,
    name: d.name,
    level: d.level,
    description: d.description,
    instance_count: d.instance_ids.length,
    created_at: 'fixed',
    dataset_revision: version,
    evaluator_revision: 'official-docker-test-runner-v1',
  }));
}

export function getFixedSuite(suiteId: string): SuiteManifest {
  const def = loadFixedSuiteDefinitions().find((d) => d.suite_id === suiteId);
  if (!def) throw new Error(`fixed suite not found: ${suiteId}`);
  return generateSuite(def.level, 0);
}

export function generateSuite(
  level: SuiteLevel,
  _seed?: number | null,
): SuiteManifest {
  const instances = loadSuiteInstances();
  const byId = new Map(instances.map((i) => [i.instance_id, i]));
  const defs = loadFixedSuiteDefinitions();
  const def = defs.find((d) => d.level === level);
  if (!def) {
    throw new Error(`未找到固定套件定义: ${level};请在 src/data/fixed_suites.json 中维护`);
  }

  const chosen: Instance[] = [];
  for (const id of def.instance_ids) {
    const inst = byId.get(id);
    if (!inst) {
      throw new Error(`固定套件 ${def.suite_id} 引用了不存在的实例: ${id}`);
    }
    chosen.push(inst);
  }

  const annotated = new Map(
    difficulty.annotate(chosen).map((r) => [r.instance_id as string, r]),
  );
  const quotas: Record<string, number> = { easy: 0, medium: 0, hard: 0 };
  for (const i of chosen) {
    const band = difficulty.difficultyBand(i);
    quotas[band] = (quotas[band] || 0) + 1;
  }

  const now = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
  const meta = OFFICIAL_META as Record<string, unknown>;
  const manifest: SuiteManifest = {
    suite_id: def.suite_id,
    suite_version: `sbp-fixed-${def.suite_id}`,
    level,
    seed: 0,
    dataset_revision: String(meta.version ?? 'official-v1'),
    evaluator_revision: 'official-docker-test-runner-v1',
    scaffold_revision: 'agent-scaffold-v1',
    created_at: now,
    quotas: { primary: quotas },
    relaxations: ['固定套件：不随机抽样，题目由 src/data/fixed_suites.json 固定'],
    instances: chosen.map((i) => {
      const row = annotated.get(i.instance_id)!;
      return {
        ...row,
        problem_statement: i.problem_statement,
        requirements: i.requirements,
        interface: i.interface,
        fail_to_pass: i.fail_to_pass,
        pass_to_pass: i.pass_to_pass,
        // 评测必需字段：没有 test_patch 则 F2P 测试无从注入，仓库/安装信息缺失会导致评测跑错目录
        base_commit: i.base_commit,
        repo_directory: i.repo_directory,
        install: i.install,
        test_cmd: i.test_cmd,
        test_patch: i.test_patch,
      };
    }),
  };
  return manifest;
}

File: app/sampler.py
"""分层抽样器(设计文档 §6 / §8 / §9)。

按难度/任务类型/语言/仓库配额建立槽位,每轮从候选池选择
InfoScore + diversity_gain 最高且满足约束的任务;
配额无法满足时按 仓库上限 → 相邻难度带 的顺序放宽并记录。
"""
from __future__ import annotations

import secrets
from datetime import datetime, timezone

from . import dataset, difficulty
from .schemas import Instance, SuiteLevel, SuiteManifest

QUOTAS: dict[str, dict[str, int]] = {
    "smoke6": {"easy": 2, "medium": 2, "hard": 2},
    "core12": {"easy": 2, "medium": 6, "hard": 4},
    # confirm24 = core12 基础上再补 12 题(文档 §8.3:中等难度仍占一半左右)
    "confirm24_addon": {"easy": 2, "medium": 6, "hard": 4},
}

REPO_MAX = {"smoke6": 1, "core12": 2, "confirm24": 2}

# Core-12 约束:至少覆盖的仓库数 / 每类任务最低数量(文档 §8.2)
TYPE_FLOOR = {"bug": 3, "feature": 3, "refactor": 2, "infra": 2}


def _diversity_gain(inst: Instance, selected: list[Instance]) -> float:
    """奖励尚未覆盖的语言/仓库/任务类型/知识域(4 维命中率)。"""
    dims = ["language_family", "task_type", "repo", "knowledge_domain"]
    fresh = 0
    for dim in dims:
        covered = {getattr(s, dim) for s in selected}
        if getattr(inst, dim) not in covered:
            fresh += 1
    return fresh / len(dims)


def _select_band(candidates: list[Instance], band: str, selected: list[Instance],
                 repo_count: dict[str, int], repo_max: int,
                 type_needs: dict[str, int]) -> tuple[Optional[Instance], list[str]]:
    """在一个难度带内选择最优候选(InfoScore + 多样性,平分按 instance_id)。"""
    relaxations: list[str] = []
    pool = [c for c in candidates if difficulty.difficulty_band(c) == band]

    def eligible(c: Instance, relax_repo: bool) -> bool:
        if repo_count.get(c.repo, 0) >= repo_max and not relax_repo:
            return False
        return True

    def score(c: Instance) -> float:
        band_p = difficulty.band_pass_prob(band)
        # 未满足的任务类型配额给予额外加分,推动类型覆盖
        type_bonus = 0.15 if type_needs.get(c.task_type, 0) > 0 else 0.0
        return difficulty.info_score(c, band_p, _diversity_gain(c, selected)) + type_bonus

    for relax_repo in (False, True):
        avail = [c for c in pool if eligible(c, relax_repo)]
        if relax_repo and avail:
            relaxations.append(f"{band}: 仓库上限约束放宽(候选不足)")
        if avail:
            return max(avail, key=lambda c: (score(c), c.instance_id)), relaxations
    return None, relaxations


def _fill_quotas(candidates: list[Instance], quotas: dict[str, int],
                 selected: list[Instance], repo_count: dict[str, int],
                 repo_max: int, base_selected: list[Instance] | None = None,
                 base_repo_count: dict[str, int] | None = None) -> list[str]:
    """按配额贪心填充;难度带候选枯竭时向相邻带漂移。"""
    relaxations: list[str] = []
    taken_ids = {i.instance_id for i in selected}

    def pick_band(band: str) -> Optional[Instance]:
        base = base_selected or []
        base_rc = base_repo_count or {}
        type_needs = dict(TYPE_FLOOR)
        for i in selected:
            if i.task_type in type_needs and type_needs[i.task_type] > 0:
                type_needs[i.task_type] -= 1
        inst, rel = _select_band(
            [c for c in candidates if c.instance_id not in taken_ids],
            band, base + selected,
            {**base_rc, **repo_count}, repo_max, type_needs,
        )
        relaxations.extend(rel)
        return inst

    # 每轮优先补数量缺口最大的带,保证带间均衡
    remaining = dict(quotas)
    drift_order = {"easy": ["medium"], "medium": ["easy", "hard"], "hard": ["medium"]}
    while any(v > 0 for v in remaining.values()):
        band = max((b for b in remaining if remaining[b] > 0),
                   key=lambda b: (remaining[b], b))
        inst = pick_band(band)
        if inst is None:
            # 本带候选枯竭:从配额已满的相邻带漂移补位
            for alt in drift_order.get(band, []):
                if remaining.get(alt, 0) > 0:
                    continue
                inst = pick_band(alt)
                if inst is not None:
                    relaxations.append(
                        f"{band} 候选池耗尽,由相邻带 {alt} 漂移补位({inst.instance_id})")
                    break
            if inst is None:
                relaxations.append(f"{band} 带配额缺口 {remaining[band]} 无法满足(候选池不足)")
                remaining[band] = 0
                continue
        selected.append(inst)
        taken_ids.add(inst.instance_id)
        repo_count[inst.repo] = repo_count.get(inst.repo, 0) + 1
        remaining[band] -= 1
    return relaxations


def generate_suite(level: SuiteLevel, seed: int | None = None) -> SuiteManifest:
    """生成套件清单。抽样完全确定性(由配额+分数决定),seed 记录进版本指纹。"""
    instances = dataset.load_instances()
    seed = seed if seed is not None else secrets.randbelow(10_000)
    annotated = {r["instance_id"]: r for r in difficulty.annotate(instances)}

    if level == "smoke6":
        selected: list[Instance] = []
        repo_count: dict[str, int] = {}
        relax = _fill_quotas(instances, QUOTAS["smoke6"], selected, repo_count,
                             REPO_MAX["smoke6"])
        chosen = selected
    else:
        # Core-12 作为 Confirm-24 的基础(同一 level 内自洽)
        core: list[Instance] = []
        repo_count: dict[str, int] = {}
        relax = _fill_quotas(instances, QUOTAS["core12"], core, repo_count,
                             REPO_MAX["core12"])
        chosen = core
        if level == "confirm24":
            addon: list[Instance] = []
            relax += _fill_quotas(instances, QUOTAS["confirm24_addon"], addon,
                                  repo_count, REPO_MAX["confirm24"],
                                  base_selected=core, base_repo_count={
                                      r: repo_count.get(r, 0) for r in {i.repo for i in core}})
            chosen = core + addon

    now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    version = f"sbp-mini-{datetime.now().strftime('%Y.%m')}-{level}-s{seed}"
    manifest = SuiteManifest(
        suite_id=f"suite-{level}-s{seed}",
        suite_version=version,
        level=level,
        seed=seed,
        dataset_revision=dataset.SEED_META.get("version", "seed-demo-v1"),
        evaluator_revision="builtin-mock/heuristic-v1",
        scaffold_revision="single-turn-patch-scaffold-v1",
        created_at=now,
        quotas={"primary": QUOTAS["smoke6" if level == "smoke6" else "core12"],
                **({"addon": QUOTAS["confirm24_addon"]} if level == "confirm24" else {})},
        relaxations=relax,
        instances=[dict(annotated[i.instance_id],
                        problem_statement=i.problem_statement,
                        requirements=i.requirements,
                        interface=i.interface,
                        fail_to_pass=i.fail_to_pass,
                        pass_to_pass=i.pass_to_pass)
                   for i in chosen],
    )
    return manifest
